lighttable rewrite

AlicVB opened the pull request:
So it's time to open my lighttable rewrite for testings.

The main goal of this work is to avoid the "draw all thumbs in one unique widget" in favor of one specific widget for each thumb (and even more 1 specific widget for each part of the thumb)

So the rewrite is mainly on code level, and there shouldn't be lot of ui change (and most of them can be tweaked by css now)

### Technical changes :
- filemanager/zoomable/filmstrip are now the same widget, so they should share the same features and work the same way.
- as widgets are separated, that should save lot of cpu cycles, esp. avoid lot of db request.
- and now every bits of thumbnails can be tweaked by css

### filmstrip :
- in order to preserve selection from unwanted modification, only click with modifiers keys can change it when in filmstrip. So in culling, you can activate another image from the filmstrip, without loosing what you have selected
- as a consequence of this, filmstrip use now simple click to switch to another image (already the case for culling/preview, only change is for darkroom)

### offset modifications :
In case of image modification which change images collection (ex : rating change). Here's a copy of the code comment which explain how offset is changed :
```
/** Here's how it works
*
*          list of change|   | x | x | x | x |
*  offset inside the list| ? |   | x | x | x |
* offset rowid as changed| ? | ? |   | x | x |
*     next imgid is valid| ? | ? | ? |   | x |
*                        |   |   |   |   |   |
*                        | S | S | S | S | N |
* S = same imgid as offset
* N = next untouched imgid as offset
**/
```

### images to act on :
On which images do we apply changes ?
```
/** Here's how it works
*
*             mouse over| x | x | x |   |   |
*     mouse inside table| x | x |   |   |   |
* mouse inside selection| x |   |   |   |   |
*          active images| ? | ? | x |   | x |
*                       |   |   |   |   |   |
*                       | S | O | O | S | A |
*  S = selection ; O = mouseover ; A = active images
*  the mouse can be outside thumbtable in case of filmstrip + mouse in center widget
*  active images are the one shown in the center part of the screen (culling / preview / darkroom)
**/
```
The idea is to centralize all that in one unique place, so if one want to propose a different algorithm as an option, it should be easier.
In this PR, this has only been done for actions directly accessible from filemanager accels/buttons : ratings/colorlabels/history copy/...

### log messages :
- log messages are now display on a specific widget on top of the center widget area.
- this was required for filemanager/zoomable
- this should solve lot of glitch, esp. when switching views/layouts
- this has allowed me to add a feature : clicking on the log msg hide it !

### What has not been done ?
In order to keep this huge PR as small as possible, I've let some things for futures PR :
- Use same "specific widgets" logic for culling and preview (should be far easier and less intrusive)
- expand the "images to act on" logic to all libs (and handle correctly grouped images)


### What to test :
Considering the amount of testing/preview requests, I propose this with only minimal testings. So you have to expect lot of bugs/crashs/glitchs/...
- offsets images when switching views / changing collection / ...
- visual glitchs. Note that for css, I would prefer to wait for @Nilvus review !
- regressions from actual behaviors
- ...

@TurboGit  : I know the amount of work I'm leaving to you for the review... If I can do anything to simplify that, just ask (rebasing, direct discussion in French, ...)
@Nilvus  , when you have some time, can you have a look at the css ? (I've tried to write lot of explanations in the css file) I can change things if you need different approach. Esp. I don't know how you want to define colors (currently, I use a weird mix...). I've not touch the other themes, maybe better stabilize this one before anything !

Nilvus commented:
@AlicVB: as the weather is coming to be quite rainy for the next days (including a rainy week-end), I should have time to test that probably on saturday or sunday, and of course especially CSS part.

phweyland commented:
Wow ! Promising !
Just gone through quickly. I like the speed ! And the single click to open an image in darkroom !
So far I just find image information when hovering not readable enough and a bit invasive (lighttable). 
Great work !

AlicVB commented:
@phweyland : what information are you referring too ? nothing should have really changed in this aspect, I've tried to mimic the old behavior. Can you share a screenshot, maybe ?
Thanks :)

TurboGit commented:
Just a quick test, I found that when the thumbs are not taking all the spaces in the height (4 rows or 11 row on my side) scrolling down does stop working. The ctrl-scroll to change the number of thumbs per row does not seem to work.

On the good side the speed is impressive especially visible when using the zoomable lighttable!

TurboGit commented:
And BTW, you need to rebase and fix conflict due to the merging of the pipeline fix.

TurboGit commented:
The single click on the filmstrip does not play well with the following scenario:
- on darkroom edit an image
- copy the dev
- paste on some other images needed the same iop

You cannot do that without editing the first image which is automatically selected, so you need to select back the originally edited image after pasting.

TurboGit commented:
And BTW a single click on the filmstrip does not deselect previously selected pictures.

phweyland commented:
> Can you share a screenshot, maybe ?

sure.

new version (hovering)
![image](https://user-images.githubusercontent.com/23012047/75389313-1574d000-58c5-11ea-8369-b4461ab52ee6.png)

previous version (hovering)
![image](https://user-images.githubusercontent.com/23012047/75389413-3806e900-58c5-11ea-9a1c-b233fd641420.png)

new version (hovering and hovering overlay)
![image](https://user-images.githubusercontent.com/23012047/75390769-9a60e900-58c7-11ea-87e0-664c3b6106b1.png)


previous version (hovering but static overlay)
![image](https://user-images.githubusercontent.com/23012047/75390697-77363980-58c7-11ea-8963-758f92a60c3a.png)

AlicVB commented:
> I found that when the thumbs are not taking all the spaces in the height (4 rows or 11 row on my side) scrolling down does stop working.

Not sure to understand... apart for some very rare cases you should always have "not fully visible" thumbs on the bottom of your page. It's not the case for you ? (or I don't understand "taking all the spaces in the height")

> The ctrl-scroll to change the number of thumbs per row does not seem to work.

oops, forgotten feature. Should be fixed now

for the filmstrip part, I fear that what you report is "on purpose" : 
- the idea is to not loose your selection when entering layouts with filmstrip, (filmstrip is not really suited to play with selections) except if you really know what you are doing and then you have to use modifier keys (ctrl / shift) Note that I've tried to add alt-click but for some reasons alt modifier is not handled at all :(
- for the copy paste, you have to hover in filmstrip the destination image and paste. You don't need to click on it.
I've done all that to be consistent with the behavior of other dt parts (but I'm all open to discuss that, as usual)

@phweyland : it's defined in css. Maybe try to lower transparency with something like : 
```
.dt_extended_overlay.dt_show_overlays #thumb_bottom
{
  background-image: linear-gradient(0deg, rgba(220, 220, 220, 0.95) 0%, rgba(220, 220, 220, 0.95) 95%,rgba(220, 220, 220, 0) 100%);
}
```

TurboGit commented:
> Not sure to understand... apart for some very rare cases you should always have "not fully visible" thumbs on the bottom of your page. It's not the case for you ? (or I don't understand "taking all the spaces in the height")

I understand that I was not clear, I feared about that, let me try again.

On the lt you have a set of thumbs displayed. I found that I was not able to scroll through all images using the mouse in some cases. Using the scrollbar always work. The issue is when scrolling to see more thumbs that are at the bottom of the lt.

What I have observed is that if the last line on the lt is fully displayed and does take all the spaces at the bottom (bottom of each thumbs on this line are exactly touching the bottom of the lt) the scrolling does not work and I cannot show more thumbs, in other words I cannot move up the current content to let some more thumbs appear at the bottom.

If the last line is having part of the thumbs displayed (for example 50% of the height of the thumbs are visible), all seems fine. That's why depending on the thumbs size (controlled by the number of thumbs per line) the issue is present or not. On my side I had the issue with 4 thumbs per line, but this really depends on your screen size and lt layout.

Hope this clarify the situation.

TurboGit commented:
>     * (filmstrip is not really suited to play with selections) except if you really know what you are doing

I find this strange to say this. This is the only way to apply a set of iop to images. When I'm working on a set of picture I'm first working on a image for basic things like exposure, filmic, noise reduction... and when I'm happy I apply the same dev to all images taken in the same context. It would be nice to have to go back to lighttable for that and back to darkroom. So to me having a way to control selection from filmstrip in the darkroom is a key feature.

johnny-bit commented:
> I apply the same dev to all images taken in the same context

You do that via `ctrl-shift-c ctrl-shift-v` keys? I think this workflow needs to be a bit better documented as it really speeds up img dev time.

TurboGit commented:
@johnny-bit : Yes indeed, I'm doing that with `ctrl-shift-[c v].`

AlicVB commented:
> Hope this clarify the situation.

Yes, I got it now, sorry for my bad English. I'll try and fix this evening...

> It wouldN'T be nice to have to go back to lighttable for that and back to darkroom. So to me having a way to control selection from filmstrip in the darkroom is a key feature.

agree. What I wanted to avoid is the modification of the selection by simple click only, otherwise you can't keep a selection and switch to another image in darkroom.
Now your use case is perfectly ok (and I use it from time to time). That's why modification of the selection is still possible, but just require modifiers keys : 
- Ctrl + click = toggle the image selection
- Shift + click = range selection
- Alt + click = select only the image (and here is the problem I have to solve : that doesn't work currently as Alt modifier doesn't seems to be recognized by gtk, no idea why... alt-gr works...)
- I wanted to add ctrl-A ctrl-shift-A too, but it was complex to avoid conflicts with already existing accels keys

TurboGit commented:
Ok, I'll try that. And yes indeed I also do you ctrl-a to select all images to apply the same iop, most of the case on many images (300 or more) from a show where I need to apply denoise profile on all pictures (and again after having fine tuned the correction on the first image).

AlicVB commented:
Thanks for your tests...
- Alt-click should work now, apart if you have a strange wm like me (xfce) which use this combinaison for something else...
- Ctrl-A should work too
- the bug which prevent to scroll down in certain cases should be fixed too

TurboGit commented:
Thanks, the scrolling issue seems fixed indeed. Some more issues:
- from darkroom
- copy one module (I'm using monochrome as result is easy to see)
  `ctrl-shift-c`
- paste on an image (overring on it)
  `ctrl-v`

=> nothing happens, if you click on this image the thumbs is refreshed and indeed monochrome has been pasted. Probably a missing refresh.

After that I did over a group of selected images, `ctrl-v` again nothing displayed, I entered the first image ok monochrome appears, I clicked on the second and got a crash. Maybe something to watch here... No clean reproducer yet.

TurboGit commented:
Playing around with culling (`ctrl-x` with 3 images selected) mode back and forth to lighttable I got:

`darktable: ../../../../src/cairo-surface.c :955 : cairo_surface_destroy:  l'assertion « CAIRO_REFERENCE_COUNT_HAS_REFERENCE (&surface->ref_count) » a échoué.`

TurboGit commented:
Ok, one reproducer:
- from lighttable enter culling mode `x`
- zoom in one image
- pan a bit
- scroll to go to next image
- leave culling
- enter it again
- leave it and dt crash

AlicVB commented:
@TurboGit : 
- the Ctrl-V refreshing is fixed
- I haven't reproduced the crash after Ctrl-V in multiple images (maybe the precedent fix, fixed that too ?)
- I haven't reproduced the bug with culling, but seeing your report, I've added a check on the thumbnail destroy code. Can you test if it solves the problem ?
- (not related to one of your reports) Now zooming in/out should keep the right image visible on screen, even with accels/slider

TurboGit commented:
Ok, `ctrl-v` working now.

I can still reproduce the culling issue, looks like you need to move the mouse over the lighttable when leaving culling. It reproduce quite often for me and always with the scenario I described above.

TurboGit commented:
If it can help, this comes from a signal and pass by the routine _event_image_draw():

```
#1  0x00007ffff7a7a535 in __GI_abort () at abort.c:79
#2  0x00007ffff7a7a40f in __assert_fail_base
    (fmt=0x7fffeed468d3 "%s%s%s :%u : %s%s l'assertion « %s » a échoué.\n%n", assertion=0x7ffff6ef8298 "CAIRO_REFERENCE_COUNT_HAS_REFERENCE (&surface->ref_count)", file=0x7ffff6ef8250 "../../../../src/cairo-surface.c", line=930, function=<optimized out>) at assert.c:92
#3  0x00007ffff7a87b92 in __GI___assert_fail
    (assertion=0x7ffff6ef8298 "CAIRO_REFERENCE_COUNT_HAS_REFERENCE (&surface->ref_count)", file=0x7ffff6ef8250 "../../../../src/cairo-surface.c", line=930, function=0x7ffff6ef8810 "cairo_surface_reference") at assert.c:101
#4  0x00007ffff6e88f30 in cairo_surface_get_reference_count ()
    at /usr/lib/x86_64-linux-gnu/libcairo.so.2
#5  0x00007ffff6e63aa5 in  () at /usr/lib/x86_64-linux-gnu/libcairo.so.2
#6  0x00007ffff6e63d70 in cairo_pattern_create_for_surface ()
    at /usr/lib/x86_64-linux-gnu/libcairo.so.2
#7  0x00007ffff6e3bfd1 in  () at /usr/lib/x86_64-linux-gnu/libcairo.so.2
#8  0x00007ffff6e96617 in cairo_set_source_surface ()
    at /usr/lib/x86_64-linux-gnu/libcairo.so.2
#9  0x00007ffff7e087aa in _event_image_draw
    (user_data=0x555558a73000, cr=0x55555780e200, widget=0x555558798450)
    at /home/obry/dev/builds/darktable/build/src/src/dtgtk/thumbnail.c:231
#10 _event_image_draw
    (widget=0x555558798450, cr=0x55555780e200, user_data=0x555558a73000)
    at /home/obry/dev/builds/darktable/build/src/src/dtgtk/thumbnail.c:110
#11 0x00007ffff745c4bb in  () at /usr/lib/x86_64-linux-gnu/libgtk-3.so.0
#12 0x00007ffff740651f in  () at /usr/lib/x86_64-linux-gnu/libgtk-3.so.0
#13 0x00007ffff6bcf0e6 in  () at /usr/lib/x86_64-linux-gnu/libgobject-2.0.so.0
#14 0x00007ffff6beac3b in g_signal_emit_valist ()
    at /usr/lib/x86_64-linux-gnu/libgobject-2.0.so.0
#15 0x00007ffff6beb81f in g_signal_emit ()
    at /usr/lib/x86_64-linux-gnu/libgobject-2.0.so.0

```

TurboGit commented:
Clicking anywhere on the lighttable canvas was clearing the selection (no image selected). This was quite handy.

TurboGit commented:
For the culling view I cannot reproduce when a filmroll of 3 pictures. Maybe this needs to have lot of thumbnails to reproduce? I can reproduce on a filmroll of 18 pictures (and one with 97).

Nilvus commented:
@AlicVB: a really awesome work you made here. A more tweakable lighttable and quite more faster also. Thanks a lot for this great work.

Another issue: I can't apply color labels with shortcuts ones or even with color buttons on the bottom left. And I can apply star ratings directly on thumbnails but not with those on the bottom left.

About color labels, a glitch on images I have with a red or yellow color labels that are applied partially on text line above. See below:

![Capture-20200229000656-1442x874](https://user-images.githubusercontent.com/45535283/75604773-ba1d2a80-5adc-11ea-9b95-7461566dbb00.png)

Nilvus commented:
Other thing that would be good to improved is text image on culling mode.
When you use two images on landscapes format, text is on the top left of the image (inside the image). When you use two images on portrait format or two images, one in landscape, one in portrait, the text is partially or totally outside the image in the top left. Not really consistent and sometimes quite awful. I don't know if it could be easily fix or not, but that would be good.

If needed, I could post screen captures.

Nilvus commented:
@AlicVB: another issue, Ctrl+Z (probably also Ctrl+Y) doesn't work. Test I just made is copy/paste a development from an image to another. I wanted to cancel the paste with Ctrl+Z but nothing happen.

AlicVB commented:
Great ! some more work for this rainy day...
First the display issues : 
> About color labels, a glitch on images I have with a red or yellow color labels that are applied partially on text line above. See below:

Yep, text need a little more space under... I'll fix that

> Another thing (not a really annoying one as only developers will need that): to work on CSS, I use quite often the Gtk Inspector Tool. The shortcut to launch it is to press 'Shift+Ctrl+D' but with your rework, this create a clone of the selected image. I have to open the about dialog window to launch Gtk Inspector Tool and not clone an image. I never had this behavior with master.

Yes, I've faced that too. Seems that it's due to how 'Shift+Ctrl+D' is handled in the code. I'm reluctant to switch back to use the key_press routines, but as a workaround (more simple than yours) you can just click on the title bar of dt and then 'Shift+Ctrl+D'. Note that dt isn't the only app around with this problem...

> Other thing that would be good to improved is text image on culling mode.

As said, I've not touch culling or preview for instance. I plan to do that once this PR will be merged : that reduce a little the size of changes and I'll have a solid base to develop that !

For the other issues, I'll try hard to reproduce, fix and push !

Nilvus commented:
> Yes, I've faced that too. Seems that it's due to how 'Shift+Ctrl+D' is handled in the code. I'm reluctant to switch back to use the key_press routines, but as a workaround (more simple than yours) you can just click on the title bar of dt and then 'Shift+Ctrl+D'. Note that dt isn't the only app around with this problem...

Ok, so forgot that. Strange that I never had that with master and I've tried to click on title bar before posting but visibly too quick... That's ok now.

> > Other thing that would be good to improved is text image on culling mode.
> 
> As said, I've not touch culling or preview for instance. I plan to do that once this PR will be merged : that reduce a little the size of changes and I'll have a solid base to develop that !
> 
> For the other issues, I'll try hard to reproduce, fix and push !

Just a thing to add about color labels. Applying them seems in realty to work but the color label is not updated. I don't try actually if it the same for star ratings but probably is.

I have seen the real behavior just a few minutes ago. I'm working on lot of images and apply the green color when an image is finished. So, to see which images I have to work on. I apply filter and exclude green ones. I have finish an image and apply the green label and the image disappear. Tested by clicking on the bottom left icons and with the shortcut. I have tested on other images with a yellow one. The color label doesn't appear on the image thumbnail but when I go to the filter option, I see that the image is on yellow filter. If I quit the actual collection and go back again, the color label then appear on the thumbnail.

So that's seems to be more a refresh display of color labels (and probably star ratings) on thumbnails than anything else.

Hope this helps you !

Nilvus commented:
@AlicVB: another thing, I have 2 times darktable crashing (when coming back to lighttable from darkroom and using a shortcut on lighttable just after. That was not the same shortcut so not easy to reproduce). Anyway, I have had on Terminal these error lines:

```
darktable: ../../../../src/cairo-surface.c :955 : cairo_surface_destroy:  l'assertion « CAIRO_REFERENCE_COUNT_HAS_REFERENCE (&surface->ref_count) » a échoué.
Abandon
```

```
darktable: ../../../../src/cairo-surface.c :930 : cairo_surface_reference:  l'assertion « CAIRO_REFERENCE_COUNT_HAS_REFERENCE (&surface->ref_count) » a échoué.
Abandon
```

AlicVB commented:
@Nilvus : yep, I'm on it currently, seems like the same problem @TurboGit report. I think we need a mutex somewhere. I'm investingating. and here, it's more easy to reproduce if you have heavy modules on (ex : profile denoise with high radius)

AlicVB commented:
for the crash with `cairo_surface` reported by @TurboGit and @Nilvus : 
They should be fixed now (hopefully as I can't reproduce exactly the culling one...)

for the Ctrl-Z : 
It should be fixed, it was just a lack of refreshing. That said, I have a question on how undo works : 
- in `dt_undo_item_t`, you have an `undo` routine that accept a list of imgid as last argument (`imgs`)
- this list seems sometimes to be filled with the imgid of changed images, sometimes not
- esp, `dt_history_snapshot_undo_pop` doesn't record anything to the list, contrary to `_pop_undo` in `rating.c` for example.
- This list can be handy to specify which images have changed when updating query.
- => Is there any reason to not record the list ?

AlicVB commented:
> Clicking anywhere on the lighttable canvas was clearing the selection (no image selected). This was quite handy.

hmm, you means when you are in culling mode, right ? If so, I can have a look, but I'm not sure it's a good idea : we try hard to avoid that user loose their selection "by accident" and I fear that clicking anywhere in dt canvas is exactly the case to avoid... Doesn't Ctrl-Shift-A sufficient for this ? (and note that it'll not be consistent with other layouts anyway)

TurboGit commented:
> hmm, you means when you are in culling mode, right ?

No I'm speaking of the lighttable.

TurboGit commented:
Yes `ctrl-shif-a` does that. But to select you need to handle the mouse and so here you'll need to leave it, to go on keyboard and back. In fact I'm using `ctrl-a` but I have never used `ctrl-shif-a`

AlicVB commented:
@TurboGit : so it's a feature I'm not aware of :)
I've just tested with master, in lighttable filemanager mode, and I suppose that when you says "click anywhere in the canvas" you means in areas where there's no thumb (bottom part or top if first image is not at first place) ?
If it's that, I agree with you, and will add that asap !

AlicVB commented:
@TurboGit : done

TurboGit commented:
Still rainy day tomorrow :) So here is another issue...

You you change the filter on top of lightttable you'll see that the refresh is not done. Change from color label and start for example. Hover over the lt and you'll see the refresh done and see that actually the image are properly sorted just that the metadata on overlay were not refreshed..

TurboGit commented:
@AlicVB : Same issue if you change the filter.

AlicVB commented:
> You you change the filter on top of lightttable you'll see that the refresh is not done. Change from color label and start for example. Hover over the lt and you'll see the refresh done and see that actually the image are properly sorted just that the metadata on overlay were not refreshed..

and another one I can't reproduce :(
Just to be sure we are doing the same things : 
- you are in lighttable, layout filemanager.
- you change sorting order with mouse (and the context menu proposing the orders overlap the central view)
- the first image should stay the same, but the other may be different, depending of their ratings/colorlabels/... values.

But that not the case, you need to hover thumb with the mouse for images to refresh, and the overlays don't change.

I'm right ?

If so, it's really strange... can you run dt with `-d lighttable` and see if you have msg in console when you change sorting order (you should) ?
Thanks

TurboGit commented:
> Just to be sure we are doing the same things :
> 
>     * you are in lighttable, layout filemanager.

Yes.

>     * you change sorting order with mouse (and the context menu proposing the orders overlap the central view)

Yes.

>     * the first image should stay the same, but the other may be different, depending of their ratings/colorlabels/... values.

Yes, but the overlay (I have the overlay info always activated on the thumbs) are not refreshed. They are refreshed only when flying over with the mouse. I just suspect missing refresh events when sorting or filters are changed.

Nilvus commented:
> for the crash with `cairo_surface` reported by @TurboGit and @Nilvus :
> They should be fixed now (hopefully as I can't reproduce exactly the culling one...)
> 
> for the Ctrl-Z :
> It should be fixed, it was just a lack of refreshing. That said, I have a question on how undo works :
> 
>     * in `dt_undo_item_t`, you have an `undo` routine that accept a list of imgid as last argument (`imgs`)
> 
>     * this list seems sometimes to be filled with the imgid of changed images, sometimes not
> 
>     * esp, `dt_history_snapshot_undo_pop` doesn't record anything to the list, contrary to `_pop_undo` in `rating.c` for example.
> 
>     * This list can be handy to specify which images have changed when updating query.
> 
>     * => Is there any reason to not record the list ?

Color labels are ok now. Thanks. Ctrl+Z doesn't work by now. I've succeeded to have it work one or two times only. It's quite strange and by now I don't understand the behavior. So this is not fix.

AlicVB commented:
> I have the overlay info always activated on the thumbs

That was the part I was missing to reproduce. Should be fixed now !

I've fixed a bug with undo if you click on star inside the thumb too.
Now seeing your edit, I'll do some more tests because I fear that there's another problem.

AlicVB commented:
> the refreshing problem about Ctrl+Z is here only when you have multiple conditions on filter collections. If I have only one default condition, Ctrl+Z is ok but if, for example, I use film roll, then add a condition to exclude green labels, then adding a green label to an image make it disappear (that's wanted) but if I press Ctrl+Z to cancel that, image doesn't come back. But If I reset filter and go where the image is, I see that the green label have been correctly removed.

As usual, I can't reproduce this one :( By any chance, do you have clicked on a star inside a thumb before applying colorlabel ? In this case, my last commit should have fixed the bug, otherwise I'll have to try more ;)

Nilvus commented:
> As usual, I can't reproduce this one :( By any chance, do you have clicked on a star inside a thumb before applying colorlabel ? In this case, my last commit should have fixed the bug, otherwise I'll have to try more ;)

As I remember, no. I will try to make more tests if I have time these next days.

Nilvus commented:
> > the refreshing problem about Ctrl+Z is here only when you have multiple conditions on filter collections. If I have only one default condition, Ctrl+Z is ok but if, for example, I use film roll, then add a condition to exclude green labels, then adding a green label to an image make it disappear (that's wanted) but if I press Ctrl+Z to cancel that, image doesn't come back. But If I reset filter and go where the image is, I see that the green label have been correctly removed.
> 
> As usual, I can't reproduce this one :( By any chance, do you have clicked on a star inside a thumb before applying colorlabel ? In this case, my last commit should have fixed the bug, otherwise I'll have to try more ;)

Can't understand why but indeed, can't reproduce now. After some more tests, this works... Good to me, I think this could be merged (if ok for others of course) in master. Lighttable is now really faster on my old CPU now, and displaying images is faster, smoother. Thanks again !

AlicVB commented:
@Nilvus : great ! Thanks for your tests

Nilvus commented:
> @Nilvus : great ! Thanks for your tests

Just a thing: as you know, CSS themes (grey one especially, as I see by now, it seems it will be the only one and for few things) needs to be adapted. I suggest to do that when this is merged (will be easier for me as I use mainly master branch). I should be able to do that next week.

@TurboGit or @aurelienpierre: when you merged that PR, could you just ping me (when doing that, if I don't have the time to propose that immediately, I will create an issue (to inform others users that some minor issues will be there for few days).

AlicVB commented:
@Nilvus : no pb, and don't hesitate to change also the main theme. I've done that part very quickly... and ask me if you need some change in gtk too !

TurboGit commented:
@AlicVB : I'll do a review in the coming days. I'm eager to see this last version in action.

AlicVB commented:
great !
some infos if it can help you : 
This PR is intended to be followed by 2 others : 
1. culling / full preview rewrite (in the same spirit that this one)
2. rework of the `dt_view_image_to_act_on` routine to fine tune the algorithm, take grouped images in consideration, ... (and ensure it's used everywhere)

Keeping that in mind can explain : 
- ` dt_view_image_to_act_on2` which will replace `dt_view_image_to_act_on` once 2 is done
- ` dt_view_image_to_act_on2` is only used in place in direct relation with lighttable/filmstrip. For other place, it'll be in 2
- `dt_view_image_expose` and all the mess around : they will be removed once 1 is done

If you need some more interactive informations, we can use framateam or something else...

Germano0 commented:
I just tested this pull request and it fixes https://github.com/darktable-org/darktable/issues/4168
What can I do to help?

AlicVB commented:
@Germano0 : I've updated the list of linked issues, so it should be closed once this PR is merged. For the help, just test it and find bugs if you have time, as usual ;) thanks for your proposal !

TurboGit commented:
I have started the review... This will take some time :) Huge amount of work!

aurelienpierre commented:
Thumbnails look blurry on 4K display, and focus peaking doesn't find any sharpness.

AlicVB commented:
@aurelienpierre : As I don't have an hidpi display here, can you try some things for me ?
All the surface creation happen in `dt_view_image_get_surface` `view.c:1142`
Maybe you can try to add `DT_PIXEL_APPLY_DPI` for width and height in `view.c:1150`. Thought I'm not sure it'll be sufficient for focuspeaking, which happen at the end of this function...
Thanks (and sorry to bother you with that !)

aurelienpierre commented:
> Maybe you can try to add `DT_PIXEL_APPLY_DPI` for width and height in `view.c:1150`. Thought I'm not sure it'll be sufficient for focuspeaking, which happen at the end of this function...

It doesn't solve the issue. Still looking for the culprit.

> Thanks (and sorry to bother you with that !)

Don't worry.

TurboGit commented on src/dtgtk/thumbtable.c:
```diff
@@ -0,0 +1,2114 @@
+  gtk_widget_queue_draw(table->widget);
+}
+
+void dt_thumbtable_zoom_changed(dt_thumbtable_t *table, int oldzoom, int newzoom)
```
oldzoom, newzoom can be const.

TurboGit commented on src/dtgtk/thumbtable.c:
```diff
@@ -0,0 +1,2114 @@
+
+// this is called each time collected images change
+static void _dt_collection_changed_callback(gpointer instance, dt_collection_change_t query_change, gpointer imgs,
+                                            int next, gpointer user_data)
```
const int next

TurboGit commented on src/dtgtk/thumbtable.c:
```diff
@@ -0,0 +1,2114 @@
+}
+
+static void _event_dnd_get(GtkWidget *widget, GdkDragContext *context, GtkSelectionData *selection_data,
+                           guint target_type, guint time, gpointer user_data)
```
target_type, time could be const?

TurboGit commented on src/dtgtk/thumbtable.c:
```diff
@@ -0,0 +1,2114 @@
+  return table;
+}
+
+void dt_thumbtable_scrollbar_changed(dt_thumbtable_t *table, int x, int y)
```
x, y -> const

TurboGit commented on src/dtgtk/thumbtable.c:
```diff
@@ -0,0 +1,2114 @@
+  return table->offset;
+}
+// set offset and redraw if needed
+gboolean dt_thumbtable_set_offset(dt_thumbtable_t *table, int offset, gboolean redraw)
```
offset, redraw -> const

TurboGit commented on src/dtgtk/thumbtable.c:
```diff
@@ -0,0 +1,2114 @@
+}
+
+// set offset at specific imgid and redraw if needed
+gboolean dt_thumbtable_set_offset_image(dt_thumbtable_t *table, int imgid, gboolean redraw)
```
imgid, redraw -> const

TurboGit commented on src/dtgtk/thumbtable.c:
```diff
@@ -0,0 +1,2114 @@
+                             GdkModifierType modifier, gpointer data)
+{
+  GList *imgs = dt_view_get_images_to_act_on();
+  gboolean ret = dt_history_paste_on_list(imgs, TRUE);
```
const

TurboGit commented on src/dtgtk/thumbtable.c:
```diff
@@ -0,0 +1,2114 @@
+  return dt_thumbtable_set_offset(table, _thumb_get_rowid(imgid), redraw);
+}
+
+static gboolean _accel_rate(GtkAccelGroup *accel_group, GObject *acceleratable, guint keyval,
```
keyval -> const

likewise in other instances below

TurboGit commented on src/dtgtk/thumbtable.c:
```diff
@@ -0,0 +1,2114 @@
+                                   GdkModifierType modifier, gpointer data)
+{
+  GList *imgs = dt_view_get_images_to_act_on();
+  gboolean ret = dt_history_paste_parts_on_list(imgs, TRUE);
```
const

TurboGit commented on src/dtgtk/thumbtable.c:
```diff
@@ -0,0 +1,2114 @@
+                                    GdkModifierType modifier, gpointer data)
+{
+  GList *imgs = dt_view_get_images_to_act_on();
+  gboolean ret = dt_history_delete_on_list(imgs, TRUE);
```
const

TurboGit commented on src/dtgtk/thumbtable.c:
```diff
@@ -0,0 +1,2114 @@
+  dt_accel_register_manual(NC_("accel", "view/thumbtable/rate 0"), views, GDK_KEY_0, 0);
+  dt_accel_register_manual(NC_("accel", "view/thumbtable/rate 1"), views, GDK_KEY_1, 0);
+  dt_accel_register_manual(NC_("accel", "view/thumbtable/rate 2"), views, GDK_KEY_2, 0);
+  dt_accel_register_manual(NC_("accel", "view
```
const int view

TurboGit replied:
const int rowid

TurboGit replied:
const int rowid

TurboGit replied:
cont int imgid

TurboGit replied:
const gboolean select

TurboGit replied:
const gboolean select

TurboGit replied:
const gboolean select

TurboGit commented:
@AlicVB 

Two issues found for the moment:

1. On the MAP module you cannot drag&drop a picture from the filmstrip into the map. But you can remove one geoloc by dragging a picture from the map to the filmstrip.

2. On the PRINT module, the image printed is always the one selected when entering the print module.  Go to print module, open filmstrip, select another picture. It is properly displayed on the center area. Move mouse over the picture and you'll see that the image information is for the previous picture and the filmstrip still shows the original image selected.

Let me know if something is not clear.

TurboGit commented:
Not sure if expected or not, but when using the down key to move down when we reach the last line we go to top of lighttable (that is first line). But from first line you do not go to last one with up key.

AlicVB commented on src/common/collection.c:
```diff
@@ -1830,7 +1913,24 @@ int dt_collection_image_offset(int imgid)
+  g_list_free(l);
+  return FALSE; // dont call again
+}
+static void _dt_collection_changed_callback(gpointer instance, dt_collection_change_t query_change, gpointer imgs,
```
yes, I'm not fully satisfied of this either. The problem is that `DT_SIGNAL_COLLECTION_CHANGED` has as an argument a `GList` containing the ids of affected images. We then need to free this list at some point to avoid memory leak.
iiuc `DT_SIGNAL_DEVELOP_HISTORY_WILL_CHANGE` as the same list problem too. it's solved there by freeing the list in the unique connected function.
But here, we have multiple connected functions. For instance only one used the list, but we can't be sure that it'll still be the case in the future.

So I've seen only 2 solutions for freeing the list : 
1. freeing it in `_dt_collection_changed_callback` (`thumbtable.c:997`) keeping in mind that the list we be invalid in other connecting function.
2. freeing the list in a special connected function with a long enough timeout to be reasonably sure that it'll not be in use (the list is not big in memory, so imho, it's not a problem to keep it in memory for some time)

I've not take the 1 route, as imo there's good chances that we'll need to use this list in other functions in a very near future (culling/preview, maps, ...)
For 2., I think the risk of timeout crash is really low, as it would mean that we have a connected function that takes more than 15s. to execute. and even in this case, it can be avoided by copying the list at the beginning of the function...

parafin replied:
Or you should implement correct link counting, e.g. by wrapping GList into GObject. Freeing memory on timeout like that is not something anyone should ever even think about, much less put into actual code.

TurboGit replied:
@AlicVB : Maybe a third solution (which would also "fix" the fragile implementation of DT_SIGNAL_DEVELOP_HISTORY_WILL_CHANGE). Have a way to register a cleaning routine for each signal kind. This free routine would be called when all the signals callbacks have been called. Not sure how easy this is, but it seems a generic implementation to fix this.

How does that sound?

TurboGit replied:
@AlicVB : My solution won't work as we do not control the way the signals are called. So we probably need to count the number of signal connected dt_control_signal_connect() decrement in each callback the counter and free when 0 is reached. A stupid ref counter, probably what @parafin had in mind.

But we probably also want to take into account dt_control_signal_block_by_func. Which is non trivial I suspect...

AlicVB commented on src/common/history.c:
```diff
@@ -1444,6 +1370,114 @@ const dt_history_hash_t dt_history_hash_get_status(const int32_t imgid)
   return status;
 }
 
+gboolean dt_history_copy(int imgid)
```
in fact, module order copying has been added after I've unified this. So i've just fix conflict by enabling it for the unified part... I've even not noticed the difference :(
That said, I really thing that function should behave the same if launched from lt or filmstrip. That was one of my (secondary) goal with this rewrite.
Was there any reason of this difference ?

Germano0 commented:
To stress darktable I imported 1500 photos, then I let darktable populate the lighttable view with all thumbnails, then I selected all of them and I clicked "remove" button. After that darktable replaced all thumbnails with the skull icon and it started to use a lot of CPU. I waited several minutes but the situation did not change.

I attach a GDB backtrace and the patch I used, to test the new lighttable, so we don't make confusion about which version I tested

[gdb.txt](https://github.com/darktable-org/darktable/files/4333304/gdb.txt)
[patch.txt](https://github.com/darktable-org/darktable/files/4333305/patch.txt)

johnny-bit commented:
Maybe this: https://developer.gnome.org/gobject/stable/gobject-memory.html
- it's gobject refcounting.

niedz., 15 mar 2020 o 07:55 Pascal Obry <notifications@github.com>
napisał(a):

> *@TurboGit* commented on this pull request.
> ------------------------------
>
> In src/common/collection.c
> <https://github.com/darktable-org/darktable/pull/4370#discussion_r392646893>
> :
>
> > @@ -1830,7 +1913,24 @@ int dt_collection_image_offset(int imgid)
>    return dt_collection_image_offset_with_collection(darktable.collection, imgid);
>  }
>
> -static void _dt_collection_recount_callback_1(gpointer instace, gpointer user_data)
> +static gboolean _dt_collection_free_list(gpointer user_data)
> +{
> +  if(!user_data) return FALSE;
> +  GList *l = user_data;
> +  g_list_free(l);
> +  return FALSE; // dont call again
> +}
> +static void _dt_collection_changed_callback(gpointer instance, dt_collection_change_t query_change, gpointer imgs,
>
> @AlicVB <https://github.com/AlicVB> : My solution won't work as we do not
> control the way the signals are called. So we probably need to count the
> number of signal connected dt_control_signal_connect() decrement in each
> callback the counter and free when 0 is reached. A stupid ref counter,
> probably what @parafin <https://github.com/parafin> had in mind.
>
> But we probably also want to take into account
> dt_control_signal_block_by_func. Which is non trivial I suspect...
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/darktable-org/darktable/pull/4370#discussion_r392646893>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACRRKFMRFYRIBKHGN2V52FTRHR3UTANCNFSM4K4EKWPA>
> .
>


-- 
Pozdrawiam,
Hubert Kowalski

MStraeten commented:
i tried building it on osx but it fails:
```
Scanning dependencies of target copy_history
[ 48%] Building C object src/libs/CMakeFiles/copy_history.dir/copy_history.c.o
[ 48%] Linking C shared module libcopy_history.so
undef: _dt_history_load_and_apply_on_selection
Undefined symbols for architecture x86_64:
  "_dt_history_load_and_apply_on_selection", referenced from:
      _load_button_clicked in lto.o
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [src/libs/libcopy_history.so] Error 1
make[1]: *** [src/libs/CMakeFiles/copy_history.dir/all] Error 2
make: *** [all] Error 2
```

[make verbose.txt](https://github.com/darktable-org/darktable/files/4333945/make.verbose.txt)

aurelienpierre commented:
Another one : when changing the display color profile, in light table, the thumbnails are not refreshed.

TurboGit commented:
> Maybe this: https://developer.gnome.org/gobject/stable/gobject-memory.html - it's gobject refcounting. niedz., 15 mar 2020 o 07:55 Pascal Obry <notifications@github.com> napisał(a):
> […](#)
> *@TurboGit* commented on this pull request. ------------------------------ In src/common/collection.c <[#4370 (comment)](https://github.com/darktable-org/darktable/pull/4370#discussion_r392646893)> : > @@ -1830,7 +1913,24 @@ int dt_collection_image_offset(int imgid) return dt_collection_image_offset_with_collection(darktable.collection, imgid); } -static void _dt_collection_recount_callback_1(gpointer instace, gpointer user_data) +static gboolean _dt_collection_free_list(gpointer user_data) +{ + if(!user_data) return FALSE; + GList *l = user_data; + g_list_free(l); + return FALSE; // dont call again +} +static void _dt_collection_changed_callback(gpointer instance, dt_collection_change_t query_change, gpointer imgs, @AlicVB <https://github.com/AlicVB> : My solution won't work as we do not control the way the signals are called. So we probably need to count the number of signal connected dt_control_signal_connect() decrement in each callback the counter and free when 0 is reached. A stupid ref counter, probably what @parafin <https://github.com/parafin> had in mind. But we probably also want to take into account dt_control_signal_block_by_func. Which is non trivial I suspect... — You are receiving this because you were mentioned. Reply to this email directly, view it on GitHub <[#4370 (comment)](https://github.com/darktable-org/darktable/pull/4370#discussion_r392646893)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ACRRKFMRFYRIBKHGN2V52FTRHR3UTANCNFSM4K4EKWPA> .
> -- Pozdrawiam, Hubert Kowalski

This is of no use in this context. A GList is allocated as user_data for a signal. And we just don't know how many time it will be consumed, maybe the signal won't call any callback actually or maybe all of them. A quite difficult issue to solve...

parafin commented:
As far as I can see memory should be freed by specifying destructor callback in dt_signal_description _signal_description array (signal.c file). Nothing uses this functionality right now, but it seems it exists just for that.
On the other hand it seems that glib signal system wasn't designed for this scenario at all, see https://developer.gnome.org/gobject/stable/gobject-Signals.html#signal-memory-management

TurboGit commented:
> As far as I can see memory should be freed by specifying destructor callback in dt_signal_description _signal_description array (signal.c file). Nothing uses this functionality right now, but it seems it exists just for that.
> On the other hand it seems that glib signal system wasn't designed for this scenario at all, see https://developer.gnome.org/gobject/stable/gobject-Signals.html#signal-memory-management

Right, the destructor in signal.c is connected to g_signal_connect_after, which seems what we need. 
@AlicVB : If this analysis is correct all we need is already in place here.

AlicVB commented:
@TurboGit @parafin : great ! I'll would have a look asap. Thank you very much for your help !
I'm trying to solve all the "small" other issues atm :) (only 2 left)

aurelienpierre commented:
@AlicVB I have found the culprit for blurry thumbnails : Cairo surface rescaling.

I'm writing another interpolation method, based on https://people.inf.ethz.ch/~cengizo/Files/Sig15PerceptualDownscaling.pdf and removing sRGB gamma before interpolating. While I'm at it, I might remove (at least part of) the LCMS color transform to only extract the RGB matrix and perform the matrix product with OpenMP (that will be faster).

AlicVB commented:
@aurelienpierre : great. what do you prefer, maybe we merge this first (once all the fix in place) and then you work on that ?

but that doesn't solve the focuspeaking ? right ? do you have any ideas on that ?

aurelienpierre commented:
> but that doesn't solve the focuspeaking ? right ? do you have any ideas on that ?

Focus peaking is messed up because the DPI scaling factor is applied twice it seems. It's just an issue of scaling.

> @aurelienpierre : great. what do you prefer, maybe we merge this first (once all the fix in place) and then you work on that ?

I will commit against this branch, then you can either merge my commit into this branch and update the PR, or we can push my commit in master once your work is merged. Whoever finishes his work first.

AlicVB commented:
So, I *think* that I've fixed all the small issues reported : 
- @TurboGit : maps, print, keydown
- @Germano0 : I didn't managed to reproduce the CPU charge pb, but there was a problem with remove button (and some other) when refreshing thumbs. That one is fixed, hope that it fix the CPU pb too.
- @MStraeten : `dt_history_load_and_apply_on_selection`
- @aurelienpierre : refresh after profile change

now I'll try to dig inside my signal.c mess...

AlicVB commented:
I've done a first try to fix the `DT_SIGNAL_COLLECTION_CHANGED` destruction.
@TurboGit @parafin and other, if you can have a look at https://github.com/darktable-org/darktable/pull/4370/commits/316830c7581e0c2fb083022915413ae82fa276e8

it seems to work... thought I have a small doubt (I don't know well how gtk signal work internally) : are we sure that the destruction signal is called after all connected callback function are **finished** (and not just after all connected function are **initiated**) ?

If it's good for you, I propose to do the same for `DT_SIGNAL_HISTORY_WILL_CHANGE`

parafin commented:
I don't know how glib signals work, so I'm not sure, no. But it wouldn't make much sense if _after callback doesn't wait for other callbacks to actually finish.

TurboGit commented:
Same as @parafin, if the signal is not working this way ti will be useless. So to me your patch is correct.

AlicVB commented:
I've done some naïve tests, by inserting a time consuming function at the end  of one connected signal. and the freeing is done after, so it seems that it was not so hard, after all :)
@TurboGit : to keep this PR as clean as possible, I propose to do the same for `DT_SIGNAL_HISTORY_WILL_CHANGE` but after all this is merged. What do you think ?

TurboGit commented:
@AlicVB : Works for me. I'll do another test pass and will report back.

TurboGit commented:
I found another issue, a persistent message. I have been able to reproduce 100% with those steps:
- launch dt
- select one image in lt
- go to map view
- drag a picture into the map
- select print module

You have a round box at the bottom of the view:

![Capture d’écran de 2020-03-16 17-50-51](https://user-images.githubusercontent.com/467069/76781543-2cfdf500-67af-11ea-86df-8339b96dfe98.png)


Another scenario, a bit different result:
- launch dt
- select one image in lt
- go to map view
- drag a picture into the map
- ctrl-z (undo)
- select print module

You have a persistent message "no image selected" (after more test it looks like it is a bit random, maybe last message displayed as I had also "slideshow is ready").

I think both scenario exhibit the same underlying issue.

AlicVB commented:
@TurboGit : should be fixed now

TurboGit commented on src/control/signal.c:
```diff
@@ -47,91 +47,104 @@ typedef struct dt_signal_description
+static void _collection_changed_destroy_callback(gpointer instance, int query_change, gpointer imgs,
+                                                 const int next, gpointer user_data)
+{
+  if(imgs && g_list_length(imgs) > 0)
```
Why is if(imgs) not sufficient ?

TurboGit commented:
I will merge this version as-is now. This works fine on my side. I'm pretty sure we'll have to adjust some part here and there. Better to have more field testing at this stage.

@AlicVB : Thanks a lot for this huge and important amount of work.

GLLM commented:
Hi @AlicVB 
I just want to leave a massive thank you message ... I just upgraded dt to the last available build version containing your refactoring : the rewritten lightable is miles ahead of the previous one !! :)
* thumbnails are loaded really fast
* scrolling is fast, no more stickyness
* customizable hover label : fantastic (I even discovered with de default value that there is now <b>, <i> and $(NL) .... some of which I reused in my darkroom info lines

and I'm sure I'll find even more improvements as I use it

THANKS !
Darktable is really becoming a mature & solid software ... and there are even more good stuff in the work. I'm happy !!

johnny-bit commented:
I knew that hover label from ##4441 will be liked :)

GLLM commented:
Woops @johnny-bit : very sorry about the mistake ! :-o
You can consider you are included in the people I thank above
